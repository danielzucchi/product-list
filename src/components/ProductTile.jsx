import React from 'react'
import styled from 'styled-components'
import PropTypes from 'prop-types'
import placeholderImage from '../assets/placeholder-image.png'
import LikeSign from '../assets/like.svg'

const ProductTileWrapper = styled.div`
  display: flex;
  flex-direction: column;
  width: 160px;
  margin: 10px 2px;
`

const ProductImageWrapper = styled.div`
  width: 100%;
  height: 100px;
  display: block;
  position: relative;
  &::after {
    ${({ sold }) =>
      sold &&
      `
    content: ' ';
    z-index: 1;
    display: block;
    position: absolute;
    height: 100%;
    top: 0;
    left: 0;
    right: 0;
    background: rgba(0, 0, 0, 0.5);
    `}
  }
`

const ProductImage = styled.img`
  width: 100%;
  height: 100%;
  display: block;
`

const ProductBrand = styled.p`
  margin: 0;
  font-size: 14px;
  font-weight: bold;
`

const ProductName = styled.p`
  margin: 0;
  font-size: 12px;
`

const ProductPrice = styled.p`
  margin: 0;
  font-size: 14px;
  font-weight: bold;
`

const SoldLabel = styled.p`
  color: #fff;
  font-size: 20px;
  font-weight: bold;
  text-transform: uppercase;
  text-align: center;
  margin: 0;
  padding: 0;
  position: absolute;
  top: 40px;
  right: 50px;
  z-index: 2;
`

const LikeButton = styled.button`
  background-image: url(${LikeSign});
  background-size: 15px 15px;
  background-position: center;
  background-repeat: no-repeat;
  background-color: ${({ isLiked }) => (isLiked ? '#d72d2d' : '#fff')};
  border: 1px solid #999999;
  border-radius: 5px;
  width: 30px;
  height: 30px;
  padding: 5px;
  margin: 0;
  position: absolute;
  top: 10px;
  right: 10px;
  z-index: 3;
  outline: none;
`

const addDefaultSrc = (e) => {
  e.target.src = placeholderImage
}

const ProductTile = ({
  id,
  name,
  img,
  sold,
  price,
  brand,
  isLiked,
  toggleLiked,
}) => (
  <ProductTileWrapper>
    <ProductImageWrapper sold={sold}>
      <LikeButton
        isLiked={isLiked}
        type="button"
        onClick={() => toggleLiked(id)}
      />
      {sold && <SoldLabel>Sold</SoldLabel>}
      <ProductImage src={img} onError={addDefaultSrc} alt={name} />
    </ProductImageWrapper>
    <ProductBrand>{brand}</ProductBrand>
    <ProductName>{name}</ProductName>
    <ProductPrice>{`Â£${price}`}</ProductPrice>
  </ProductTileWrapper>
)

ProductTile.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  img: PropTypes.string.isRequired,
  sold: PropTypes.bool.isRequired,
  price: PropTypes.string.isRequired,
  brand: PropTypes.string.isRequired,
  isLiked: PropTypes.bool.isRequired,
  toggleLiked: PropTypes.func.isRequired,
}

export default ProductTile
