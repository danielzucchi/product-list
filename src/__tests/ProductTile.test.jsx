import React from 'react'
import { shallow } from 'enzyme'
import ProductTile from '../components/ProductTile'

describe('ProductTile tests', () => {
  it('when all props are present matches snapshot', () => {
    const wrapper = shallow(
      <ProductTile
        id="1"
        name="Product one"
        img="img"
        sold={false}
        price="15"
        brand="sampleBrand"
        isLiked={false}
        toggleLiked={() => jest.fn()}
      />,
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('when product is sold, shows sold label', () => {
    const wrapper = shallow(
      <ProductTile
        id="1"
        name="Product one"
        img="img"
        sold
        price="15"
        brand="sampleBrand"
        isLiked={false}
        toggleLiked={() => jest.fn()}
      />,
    )

    const soldLabel = wrapper.findWhere(
      (node) => node.name() === 'styled.p' && node.text() === 'Sold',
    )

    expect(soldLabel).toHaveLength(1)
    expect(wrapper).toMatchSnapshot()
  })
})
