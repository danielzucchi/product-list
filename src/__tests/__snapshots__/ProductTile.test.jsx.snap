// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTile tests when all props are present matches snapshot 1`] = `
<styled.div>
  <styled.div
    sold={false}
  >
    <styled.button
      isLiked={false}
      onClick={[Function]}
      type="button"
    />
    <styled.img
      alt="Product one"
      onError={[Function]}
      src="img"
    />
  </styled.div>
  <styled.p>
    sampleBrand
  </styled.p>
  <styled.p>
    Product one
  </styled.p>
  <styled.p>
    £15
  </styled.p>
</styled.div>
`;

exports[`ProductTile tests when product is sold, shows sold label 1`] = `
<styled.div>
  <styled.div
    sold={true}
  >
    <styled.button
      isLiked={false}
      onClick={[Function]}
      type="button"
    />
    <styled.p>
      Sold
    </styled.p>
    <styled.img
      alt="Product one"
      onError={[Function]}
      src="img"
    />
  </styled.div>
  <styled.p>
    sampleBrand
  </styled.p>
  <styled.p>
    Product one
  </styled.p>
  <styled.p>
    £15
  </styled.p>
</styled.div>
`;
